### ISSUE TITLE
Spontaneous clone generation

### ISSUE THREAD

**Issue by tbnorth** [leo-editor/leo-editor#470]
When the same external file is present in more than one place in the outline, clones may be formed spontaneously.  This should not happen.  I created the following outline:

![clone](https://cloud.githubusercontent.com/assets/158108/24838569/e19ac972-1d0f-11e7-8a09-a23778fd5215.png)

When I created it, there were no clones.  Save, quit, re-run Leo - now there are clones.  I created the second copy of clone.py (the files are identical, I'm sure that's important) by Ctrl-Shift-C Ctrl-Shift-V.  I doubt the content of clone.py matters, but the code I used to make this example is below.

I've been seeing this in the wild for some time (over a year?) only just now making a minimal example.

```python
"""
clone.py - describe

Terry Brown, Terry_N_Brown@yahoo.com, Sun Apr  9 10:16:01 2017
"""

import argparse
import os
import sys

def make_parser():
     
     parser = argparse.ArgumentParser(
         description="""<|general description|>""",
         formatter_class=argparse.ArgumentDefaultsHelpFormatter
     )
     
     parser.add_argument("--<|foo|>", action='store_true',
         help="<|help|>"
     )
     parser.add_argument('<|positional(s)|>', type=str, nargs='+',
         help="<|help|>"
     )

     return parser
 

def main():
    opt = make_parser().parse_args()

if __name__ == '__main__':
    main()
```

**Comment by edreamleo** [comment]
Thanks for this.  I thought I had fixed this some time before.

**Comment by edreamleo** [comment]
This is an exact duplicate of #449.

**Comment by tbnorth** [comment]
Not an exact duplicate, #449 is closed, this isn't ;-)  Also, the title of #449 makes it sound very different - the auto nodes aren't cloned, just their descendants.  And this one has nothing to do with sections references.

But they're both GitHub issue reports, so definitely duplicates in that sense :-)

**Comment by edreamleo** [comment]
On Tue, Apr 11, 2017 at 7:51 PM, Terry Brown <notifications@github.com>
wrote:

> Not an exact duplicate, #449
> <https://github.com/leo-editor/leo-editor/issues/449> is closed, this
> isn't ;-) Also, the title of #449
> <https://github.com/leo-editor/leo-editor/issues/449> makes it sound very
> different - the auto nodes aren't cloned, just their descendants. And this
> one has nothing to do with sections references.
>
> But they're both GitHub issue reports, so definitely duplicates in that
> sense :-)
>

​I marked them as duplicates to remind myself of the previous fix. The same
fix, applied at a different place, will likely suffice.  Or so I say now.

Edward

**Comment by edreamleo** [comment]
Ok.  It took quite awhile to understand what you wrote :-) This behavior described happens only if:

1. Caching is enabled, that is, if the --no-cache option is _not_ in effect.
2. The "duplicate" `@auto` nodes are *not* clones of each other.  If they are, the fix for #449 works.

So yes, this is a different issue from #449. The fix for #449 has no chance of working.

A simple, good, and recommended workaround is to use clones for "duplicate" `@auto` nodes. Another workaround would be to use --no-cache when launching Leo on the .leo files in question.

The test I am using is:

```
@auto c:\test\tnb_clone_test.py
@auto c:\test\tnb_clone_test.py
@auto c:\test\tnb_clone_test.py
```
Only the last two top-level nodes are clones of each other.  Regardless of caching, the top-level nodes are shown correctly on reload: the last two nodes are always shown as clones of each other.  When caching is disabled, no child nodes are clone.  When caching is enabled, *all* child nodes are clones of each other.

The fix for this may be easy, or nearly impossible. I'll look into it.

**Comment by edreamleo** [comment]
I'm inclined not to fix this. That is, to mark this issue Won'tFix and close it.  My reasons are as follows:

1. This situation you suggest is, at best, poor practice.  It's reasonable to clone `@auto` nodes.  Using two *non-cloned* copies `@auto filename` is asking for completely unnecessary trouble. There are two apparently different files, yet they must use exactly the same source when re-loading the .leo file.

2. The code sheds some light on the situation.  The following line appears in cacher.readFile in leo.core.leoCache:

    ```key = self.fileKey(root.h, s, requireEncodedString=True)```

This computes the file's hash using *only* the `@<file>` node's headline and the contents of the file.

Suppose let the hash depend on the gnx as well? This might solve this issue, but it could cause subtle problems with caching in other situations.

In short, I am going to close this issue unless you, Terry, can give a convincing reason why your use case makes sense, and why the obvious workarounds do not suffice.

**Comment by tbnorth** [comment]
On Thu, 13 Apr 2017 06:27:19 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> I'm inclined not to fix this. That is, to mark this issue Won'tFix
> and close it.  My reasons are as follows:
> 
> 1. This situation you suggest is, at best, poor practice.  It's
> reasonable to clone `@auto` nodes.  Using two *non-cloned* copies
> `@auto filename` is asking for completely unnecessary trouble. There
> are two apparently different files, yet they must use exactly the
> same source when re-loading the .leo file.

I guess I disagree pretty strongly.  To me this is changing the normal
order of things to accommodate clones.  Having two files with
identical content on a filesystem should not be a problem for a text
editor.  

It may or may not be bad practice, depending on the context,
and even if it is bad practice in a particular case, it may not be the
Leo user's bad practice, but something the Leo user needs to be able to
deal with.

> 2. The code sheds some light on the situation.  The following line
> appears in cacher.readFile in leo.core.leoCache:
> 
>     ```key = self.fileKey(root.h, s, requireEncodedString=True)```
> 
> This computes the file's hash using *only* the `@<file>` node's
> headline and the contents of the file.
> 
> Suppose let the hash depend on the gnx as well? This might solve this
> issue, but it could cause subtle problems with caching in other
> situations.

Comfortably ignorant of all the intricacies of file loading, my answer
would be just "don't create clones without explicit instruction to do
so".  The files have the same hash, but there's no connection in my
mind between that and the creation of clones.

Because I never use clones, they're all cost and no benefit to me,
their acknowledged risks are not something I want to have to consider.

Why are they created in this case, and if there's a good reason for it,
from a clone users perspective, is it something that could be
controlled with a @setting?

Cheers -Terry


> In short, I am going to close this issue unless you, Terry, can give
> a convincing reason why your use case makes sense, and why the
> obvious workarounds do not suffice.
>

**Comment by edreamleo** [comment]
On Thu, Apr 13, 2017 at 9:54 AM, Terry Brown <notifications@github.com>
wrote:
​

> Having two files with
> ​ ​
> identical content on a filesystem should not be a problem for a text
> ​ ​
> editor.
>

​But they are not two files. They are exactly the same file.​


How do you propose that the cacher should distinguish between them? And how
do you propose that the user should distinguish between them
​?​
  What does it even mean for two separate @auto nodes (or @file or @clean
or @nosent) to refer to exactly the same path?

Edward

**Comment by tbnorth** [comment]
Gah, devilish details.  In your example they were exactly the same file, in mine, they're on different paths.  It's the case with the same file (content wise) on different paths that I think needs fixing.  The "~/clone/" vs. "~/clone2/" in the original report.

**Comment by edreamleo** [comment]
On Thu, Apr 13, 2017 at 10:04 AM, Terry Brown <notifications@github.com>
wrote:

> Gah, devilish details. In your example they were exactly the same file, in
> mine, they're on different paths. It's the case with the same file (content
> wise) on different paths that I think needs fixing. The "/clone/" vs. "/clone2/"
> in the original report.
>
​Ok.  This probably can be fixed easily and safely by having the cacher use
the fully-resolved path rather than just the node's headline.

Edward

**Comment by edreamleo** [comment]
On Thu, Apr 13, 2017 at 10:34 AM, Edward K. Ream <edreamleo@gmail.com>
wrote:

> On Thu, Apr 13, 2017 at 10:04 AM, Terry Brown <notifications@github.com>
> wrote:
>
>> Gah, devilish details. In your example they were exactly the same file,
>> in mine, they're on different paths. It's the case with the same file
>> (content wise) on different paths that I think needs fixing. The "/clone/"
>> vs. "/clone2/" in the original report.
>>
> ​Ok.  This probably can be fixed easily and safely by having the cacher
> use the fully-resolved path rather than just the node's headline.
>

​Oh joy.  Suppose the cacher prepends the git branch name (if any) to the
path? When switching branches, it will be as if --no-cache comes
automatically comes into effect.

This probably is the one and only elegant fix to #385: Leo doesn't honor
new headlines when switching branches
<https://github.com/leo-editor/leo-editor/issues/385>. This Aha is a stroke
of good fortune. I would never have made the connection without our
discussion today.

Edward

**Comment by tbnorth** [comment]
On Thu, 13 Apr 2017 09:15:46 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> On Thu, Apr 13, 2017 at 10:34 AM, Edward K. Ream <edreamleo@gmail.com>
> wrote:
> 
> > On Thu, Apr 13, 2017 at 10:04 AM, Terry Brown
> > <notifications@github.com> wrote:
> >
> >> Gah, devilish details. In your example they were exactly the same
> >> file, in mine, they're on different paths. It's the case with the
> >> same file (content wise) on different paths that I think needs
> >> fixing. The "/clone/" vs. "/clone2/" in the original report.
> >>
> > ​Ok.  This probably can be fixed easily and safely by having the
> > cacher use the fully-resolved path rather than just the node's
> > headline.
> >
> 
> ​Oh joy.  Suppose the cacher prepends the git branch name (if any) to
> the path? When switching branches, it will be as if --no-cache comes
> automatically comes into effect.

Heh - top of head that sounds very clever.  And I realize now the
getgitbranchname() method is for any path, not just Leo sources, so it
probably has no need to reference leoVersion.  But given how important
this might be - should it depend on pygit2 or should it use the method
leoVersion uses, so there's no dependency?

Cheers -Terry

> This probably is the one and only elegant fix to #385: Leo doesn't
> honor new headlines when switching branches
> <https://github.com/leo-editor/leo-editor/issues/385>. This Aha is a
> stroke of good fortune. I would never have made the connection
> without our discussion today.
> 
> Edward
> 
>

**Comment by edreamleo** [comment]
On Thu, Apr 13, 2017 at 11:21 AM, Terry Brown <notifications@github.com>
wrote:



> I realize now the
> ​ ​
> getgitbranchname() method is for any path, not just Leo sources, so it
> probably has no need to reference leoVersion. But given how important
> this might be - should it depend on pygit2 or should it use the method
> leoVersion uses, so there's no dependency?
>

​An interesting question.  My first thought was that it doesn't matter much
because g.getGitBranchName uses the leoVersion code first, and only falls
back to pygit2 in an emergency. If pygit2 doesn't exist, it returns None,
so nothing too terrible happens.

Otoh, since we are dealing with caching, this strategy adds another level
of uncertainty.  For sure, I would like to remove the dependency on pygit2
if possible.  Perhaps Leo could just adapt/steal the pygit2 code itself...

Edward

**Comment by tbnorth** [comment]
On Thu, 13 Apr 2017 09:29:46 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> On Thu, Apr 13, 2017 at 11:21 AM, Terry Brown
> <notifications@github.com> wrote:
> 
> > I realize now the ​ ​
> > getgitbranchname() method is for any path, not just Leo sources, so

Actually I see that's barely true (an alternative fact? :-}) -
g.getGitBranchName() seems to be intended for getting the Leo branch
name, although ends up being generic because it accepts a path argument.

The branch switching problem needs be be solved in general, not just
for the Leo sources.  g.getGitBranchName() can still contribute, if the
pre-pending of the branch name turns about to be useful.  Still
wondering if the *technique* leoVersion uses to get the branch name
isn't better than code that depends on a dependency, so that piece of
code could be used implement g.getGitBranchName(), which could be
called by leoVersion.

*But* I wonder if the solution should really involve git.  The problem
is that when a whole lot of files Leo has "open" change, Leo does not
handle it well.  The change could be triggered by any number of things,
a local copy, a copy from a remote source, re-extracting a .zip file,
re-running a program that updates several files.  Leo should handle all
these cases well, so a solution that requires a branch name isn't
general enough.

Using separate directories isn't a solution either, you shouldn't need
to do that and the user shouldn't need to manage that complexity.

So, this discussion is really part of #445.

As far as *this* bug is concerned, I think using the fully
qualified absolute dereferenced normalized :) path name when generating
the hash would fix it.

> > it probably has no need to reference leoVersion. But given how
> > important this might be - should it depend on pygit2 or should it
> > use the method leoVersion uses, so there's no dependency?
> 
> ​An interesting question.  My first thought was that it doesn't
> matter much because g.getGitBranchName uses the leoVersion code
> first, and only falls back to pygit2 in an emergency. If pygit2
> doesn't exist, it returns None, so nothing too terrible happens.
> 
> Otoh, since we are dealing with caching, this strategy adds another
> level of uncertainty.  For sure, I would like to remove the
> dependency on pygit2 if possible.  Perhaps Leo could just adapt/steal
> the pygit2 code itself...
> 
> Edward

**Comment by edreamleo** [comment]
On Thu, Apr 13, 2017 at 2:52 PM, Terry Brown <notifications@github.com>
wrote:

g.getGitBranchName() seems to be intended for getting the Leo branch
> name, although ends up being generic because it accepts a path argument.
>

​Heh.  It can get the branch name of any path containing a .get folder.​


>
> The branch switching problem needs be be solved in general, not just
> for the Leo sources. g.getGitBranchName() can still contribute, if the
> pre-pending of the branch name turns about to be useful.


​Yes.
​


> Still
> ​ ​
> wondering if the *technique* leoVersion uses to get the branch name
> isn't better than code that depends on a dependency, so that piece of
> code could be used implement g.getGitBranchName(), which could be
> called by leoVersion.
>

​I agree, it looks like the code in leoVersion.py should suffice in all
likely situations, and if it doesn't work, the code in pygit2 is not likely
to help.​


>
> *But* I wonder if the solution should really involve git. The problem
> ​ ​
> is that when a whole lot of files Leo has "open" change, Leo does not
> ​ ​
> handle it well.


​That's another issue (#445).
​


> The change could be triggered by any number of things,
> ​ ​
> a local copy, a copy from a remote source, re-extracting a .zip file,
> re-running a program that updates several files. Leo should handle all
> ​ ​
> these cases well, so a solution that requires a branch name isn't
> ​ ​
> general enough.
>

​Imo, caching should know about git branches, whatever the consequences.

Using separate directories isn't a solution either, you shouldn't need
> ​ ​
> to do that and the user shouldn't need to manage that complexity.
>
> So, this discussion is really part of #445.
>

​I agree.  That is the place to discuss the fine points of how the ​
ExternalFilesController should work.​

As far as *this* bug is concerned, I think using the fully
> ​ ​
> qualified absolute dereferenced normalized :) path name when generating
> ​ ​
> the hash would fix it.
>

​I agree, provided it also works with the git branch name prepended :-)

Edward

**Comment by edreamleo** [comment]
On Thu, Apr 13, 2017 at 5:12 PM, Terry Brown <notifications@github.com>
wrote:

>
>
> >​Imo, caching should know about git branches, whatever the
> >consequences.
>
> Don't see any harm, but not sure about the benefit, as a solution that
> handles the non git cases would cover git cases as well?
>

​The benefit is no new "if" statement in the caching code.

Edward

**Comment by edreamleo** [comment]
Rev b32259 fixes this issue simply by having `cacher.readFile` pass the full file name to `cacher.fileKey()`. It's surprising that passing headlines instead of file names worked at all.
